contracts/governance/Badger.sol

```diff
@@ -0,0 +1,342 @@
+     * @param tokenId       identifier for the new token tier
+     * @param uriId         identifier that is appended to the baseUri together forming the uri where the metadata lives
+     * @param transferable  determines if tokens from specific tier should be transferable or not
+     */
+    function createTokenTier(
+        uint256 tokenId,
+        string calldata uriId,
+        bool transferable
+    ) public onlyOwner isValidString(uriId) {
+        tokenTiers[tokenId] = TokenTier(uriId, transferable);
+    }
+
+    /**
+     * @dev                 updates the identifier that is appended to the baseUri for a specific tokenId (tier)
+     * @param tokenId       tokenId for which the uri should be updated
+     * @param uriId         identifier that is appended to the baseUri together forming the uri where the metadata lives
+     */
+    function updateUriIdentifier(uint256 tokenId, string calldata uriId)
+        public
+        onlyOwner
+    {
+        _updateUriIdentifier(tokenId, uriId);
+    }
+
+    /**
+     * @dev                 update uri identifiers for multiple token ids (tiers)
+     * @param tokenIds      tokenIds for which the uri should be updated (must be in same order as uriIds)
+     * @param uriIds        identifiers that are appended to the baseUri together forming the uri where the metadata lives (must be in same order ass tokenIds)
+     */
+    function updateMultipleUriIdentifiers(
+        uint256[] calldata tokenIds,
+        string[] calldata uriIds
+    ) public onlyOwner {
+        require(tokenIds.length == uriIds.length, "Input array mismatch");
+
+        for (uint8 i = 0; i < tokenIds.length; i++) {
+            _updateUriIdentifier(tokenIds[i], uriIds[i]);
+        }
+    }
+
+    /**
+     * @dev                 updates transferability for a given token id (tier)
+     * @param tokenId       tokenId for which transferability should be updated
+     * @param transferable  determines whether tokens from tier should be transferable or not
+     */
+    function updateTransferableStatus(uint256 tokenId, bool transferable)
+        public
+        onlyOwner
+        isTier(tokenId)
+    {
+        tokenTiers[tokenId].transferable = transferable;
+    }
+
+    /*
+        Queries
+    */
+
+    /**
+     * @dev                 returns the uri for a given token
+     * @notice              consists of a concatenation of baseUri and uriId
+     * @param tokenId       tokenId for which the uri should be retrieved
+     */
+    function uri(uint256 tokenId)
+        public
+        view
+        virtual
+        override
+        returns (string memory)
+    {
+        string memory baseUri = super.uri(tokenId);
+        return baseUri.append(tokenTiers[tokenId].uriId);
+    }
+
+    /*
+        Internal functions
+    */
+
+    function _updateUriIdentifier(uint256 tokenId, string calldata uriId)
+        private
+        isTier(tokenId)
+        isValidString(uriId)
+    {
```

Comment:
What's the idea here behind using `isTier` and `isValidString`?
It looks to me like both just validate that the URI string is non-empty. Wouldn't one be enough?

Reply:
Sorry, I was stupid. The idea is to check that the new uriId is valid. Ignore this :D

Reply:
They're very similar but not exactly the same. `isTier` checks whether the tier that should be updated actually exists (by looking that the `uriId` is not empty, which is a requirement for a tier) and `isValidString` checks the param, or put differently that the new `uriId` is not empty.